### PR TITLE
Add ProcessExitedError rather than using `error`

### PR DESCRIPTION
**Pull request by oxinabox**
This has been bugging me for years.
I thought it was solved during 0.7-Dev.

As someone calling external processes,
when they fail I want a unique exception type to be thrown,
so I can handle it appropriately.

I think this will require tests being updated
I wrote it as a PR because it is more clear than writing an issue I think.
But this is basically an issue.

### CODE REVIEW

File: base/process.jl [hunk 1]
```diff
@@ -259,7 +259,7 @@ run(pipeline(`update`, stdout="log.txt", append=true))
 """
 function pipeline(cmd::AbstractCmd; stdin=nothing, stdout=nothing, stderr=nothing, append::Bool=false)
     if append && stdout === nothing && stderr === nothing
-        error("append set to true, but no output redirections specified")
+        ArgumentError("append set to true, but no output redirections specified")
```
Comment (ararslan) [inline base/process.jl]: Should be `throw(ArgumentError("..."))`

File: base/process.jl [hunk 2]
```diff
@@ -707,9 +707,25 @@ with a code of 0). An exception is raised if the process cannot be started.
 """
 success(cmd::AbstractCmd) = success(_spawn(cmd))
 
+struct ProcessExitedError
+	procs::Vector{Process}
```
Comment (ararslan) [inline base/process.jl]: Four spaces for indentation, no tabs

File: base/process.jl [hunk 3]
```diff
@@ -785,7 +796,7 @@ function process_status(s::Process)
     process_signaled(s) ? "ProcessSignaled("*string(s.termsignal)*")" :
     #process_stopped(s) ? "ProcessStopped("*string(process_stop_signal(s))*")" :
     process_exited(s) ? "ProcessExited("*string(s.exitcode)*")" :
-    error("process status error")
+    @assert(false, "process status error")
```
Comment (ararslan) [inline base/process.jl]: This is kind of weird. `error` seems like a better choice here.

**Review by ararslan — CHANGES_REQUESTED** [review]

File: base/process.jl (hunk 3 above)
Comment (oxinabox) [reply base/process.jl]: assert is right if this is supposed to be a complete list of all statuss a process can have.
If not then a `UnknownProcessStatusError` or maybe a `DomainError` is suited

File: base/process.jl (hunk 2 above)
Comment (oxinabox) [reply base/process.jl]: This is what I get for not using vim.

File: base/process.jl [hunk 4]
```diff
@@ -707,9 +707,25 @@ with a code of 0). An exception is raised if the process cannot be started.
...
+        for proc in err.procs
+            println(io, "  ", proc, " [", proc.exitcode, "]")
+        end
+    end
+end
+
 function pipeline_error(proc::Process)
     if !proc.cmd.ignorestatus
-        error("failed process: ", proc, " [", proc.exitcode, "]")
+        throw(ProcessExitedError(proc)
```
Comment (ararslan) [inline base/process.jl]: Missing a close paren here

File: base/process.jl (hunk 2 above)
Comment (ararslan) [reply base/process.jl]: Looks like you got the other tabs but not this one

File: base/process.jl [hunk 5]
```diff
@@ -714,6 +714,7 @@ ProcessExitedError(proc::Process) = ProcessExitedError([proc])
 
 function show(io::IO, err::ProcessExitedError)
     if length(err.procs) == 1
+        @inbounds proc = err.procs[1]
```
Comment (KristofferC) [inline base/process.jl]: This `@inbounds` seems unnecessary.

Comment (KristofferC) [reply base/process.jl]: Because this is not in a situation where it will improve performance (nor where performance matter) and even though the length is explicitly checked above it adds extra code and if the check, by mistake, later is removed or changed, the inbounds will give silent corruption instead of an error.

Comment (oxinabox) [reply base/process.jl]: fair enough, fixed

**Comment by oxinabox** [comment]
I think I need to rebase this as I think it is based off a version of master that doesn't build

**Comment by oxinabox** [comment]
Ok testing this locally it all passes except one.
And I can't workout how I broke that one.
It doesn't seem to have anything to do with processes.

I get the error:

```
      From worker 7:    WARNING: attempting hard kill of repl test after exceeding timeout

Executing tests that run on node 1 only:
precompile                          (1) |    86.67 |   0.88 |  1.0 |    1431.78 |   278.88
SharedArrays                        (1) |    52.23 |   2.06 |  4.0 |    2085.67 |   309.06
Worker 7 failed running test REPL:
Some tests did not pass: 2 passed, 0 failed, 1 errored, 0 broken.REPL: Error During Test at /usr/src/julia/julia$
master/test/testdefs.jl:19
  Got exception LoadError("/usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/runtests.jl", 4, Loa
dError("/usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl", 74, "hard kill repl test")) o
utside of a @test
  LoadError: LoadError: "hard kill repl test"
  Stacktrace:
   [1] try_yieldto(::typeof(Base.ensure_rescheduled), ::Base.RefValue{Task}) at ./event.jl:196
   [2] wait() at ./event.jl:255
   [3] wait(::Condition) at ./event.jl:46
   [4] wait_readnb(::Base.PipeEndpoint, ::Int64) at ./stream.jl:301
   [5] eof at ./stream.jl:61 [inlined]
   [6] readuntil_vector!(::Base.PipeEndpoint, ::Base.CodeUnits{UInt8,String}, ::Bool, ::Array{UInt8,1}) at ./io.j
l:695
   [7] #readuntil#283 at ./io.jl:773 [inlined]
   [8] #readuntil at ./none:0 [inlined]
   [9] #readuntil#282(::Bool, ::Function, ::Base.PipeEndpoint, ::String) at ./io.jl:768
   [10] readuntil at ./io.jl:758 [inlined]
   [11] (::getfield(Main.Test48Main_REPL.REPLTests, Symbol("##10#15")){Base.PipeEndpoint,Base.PipeEndpoint,String
})(::String) at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl:132
   [12] mktempdir(::getfield(Main.Test48Main_REPL.REPLTests, Symbol("##10#15")){Base.PipeEndpoint,Base.PipeEndpoi
nt,String}, ::String) at ./file.jl:569
   [13] mktempdir at ./file.jl:567 [inlined]
   [14] (::getfield(Main.Test48Main_REPL.REPLTests, Symbol("##8#13")))(::Base.PipeEndpoint, ::Base.PipeEndpoint,:
:REPL.LineEditREPL) at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl:107
   [15] #fake_repl#5(::REPL.Options, ::Function, ::Any) at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.
7/REPL/test/repl.jl:53                                                                                              [16] fake_repl(::Any) at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl:42            [17] top-level scope at none:0                                                                                   [18] include at ./boot.jl:317 [inlined]                                                                          [19] include_relative(::Module, ::String) at ./loading.jl:1034                                                   [20] include at ./sysimg.jl:29 [inlined]                                                                         [21] include(::String) at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/runtests.jl:3        [22] top-level scope at none:0                                                                                   [23] include at ./boot.jl:317 [inlined]
   [24] include_relative(::Module, ::String) at ./loading.jl:1034
   [25] include at ./sysimg.jl:29 [inlined]
   [26] include(::String) at /usr/src/julia/julia-master/test/testdefs.jl:13
   [27] macro expansion at /usr/src/julia/julia-master/test/testdefs.jl:292 [inlined]
   [28] macro expansion at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/Test/src/Test.jl:1080 [inline$
]
   [29] macro expansion at /usr/src/julia/julia-master/test/testdefs.jl:21 [inlined]
   [30] macro expansion at ./util.jl:289 [inlined]
   [31] top-level scope at /usr/src/julia/julia-master/test/testdefs.jl:19 [inlined]
   [32] top-level scope at ./<missing>:0
   [33] eval at ./boot.jl:319 [inlined]
   [34] #runtests#3(::UInt128, ::Function, ::String, ::String, ::Bool) at /usr/src/julia/julia-master/test/testd$
fs.jl:25
   [35] #runtests at ./none:0 [inlined] (repeats 2 times)
   [36] (::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}})() at /usr/src/julia/jul$
a-master/usr/share/julia/stdlib/v0.7/Distributed/src/process_messages.jl:269
   [37] run_work_thunk(::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}}, ::Bool) a$
 /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/Distributed/src/process_messages.jl:56
   [38] macro expansion at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/Distributed/src/process_messa$
es.jl:270 [inlined]
   [39] (::getfield(Distributed, Symbol("##111#113")){Distributed.CallMsg{:call_fetch},Distributed.MsgHeader,Soc$
ets.TCPSocket})() at ./task.jl:257
  in expression starting at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl:74
ets.TCPSocket})() at ./task.jl:257                                                                      [219/532]
  in expression starting at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/repl.jl:74
  in expression starting at /usr/src/julia/julia-master/usr/share/julia/stdlib/v0.7/REPL/test/runtests.jl:4

```

**Comment by StefanKarpinski** [comment]
Yes, that would be great!

**Comment by oxinabox** [comment]
@ararslan did I fix everything you wanted fixing?

File: base/process.jl [hunk 6]
```diff
@@ -701,9 +701,26 @@ with a code of 0). An exception is raised if the process cannot be started.
 """
 success(cmd::AbstractCmd) = success(_spawn(cmd))
 
+struct ProcessExitedError
+    procs::Vector{Process}
+end
+ProcessExitedError(proc::Process) = ProcessExitedError([proc])
+
+function show(io::IO, err::ProcessExitedError)
```
Comment (ararslan) [inline base/process.jl]: There should be a test for this `show` method

File: base/exports.jl [hunk 7]
```diff
@@ -117,13 +117,15 @@ export
...
-    DimensionMismatch,
     CapturedException,
     CompositeException,
+    DimensionMismatch,
     EOFError,
     InvalidStateException,
     KeyError,
     MissingException,
+    ParseError,
+    ProcessExitedError,
```
Comment (ararslan) [inline base/exports.jl]: Most (but not all) other exception types have "exception" in the name instead of "error." In keeping with that, I'd call this `ProcessExitedException`, which I think sounds better anyway.

**Review by ararslan — CHANGES_REQUESTED** [review]

File: base/exports.jl (hunk 7 above)
Comment (oxinabox) [reply base/exports.jl]: I agree.
Exceptions have a mix of naming either `FooException`, or `BarError`.
I feel like we should have a convention that they are different (e.g. `Errors` are probably programmers fault. Where as who knows what causes `Exceptions`s. Or perhaps that `rrors` are normally n ot recoverable. idk).

But yet, lets name this an exception.

**Comment by oxinabox** [comment]
Problem.
The Distributed standard library exports a `ProcessExitedException` too.
I am going to rename that one to `WorkerProcessExitedException`

**Comment by ararslan** [comment]
> The Distributed standard library exports a ProcessExitedException too.

Could it not just use the one now defined in Base? That would be non-breaking.

**Comment by fredrikekre** [comment]
Download tests etc are run after the main test suite. This definitely looks related:
```
Some tests did not pass: 8 passed, 2 failed, 0 errored, 0 broken.download: Test Failed at /tmp/julia/share/julia/test/download.jl:19
  Expression: download("http://httpbin.org/status/404", missing_file)
    Expected: ErrorException
      Thrown: ProcessExitedException
```

**Comment by oxinabox** [comment]
> Download tests etc are run after the main test suite. 

Did not know that.
Do they run with `make test`? 
Or do i need to do `make testall` or some such for them?

**Comment by KristofferC** [comment]
`make test download` I think

**Comment by oxinabox** [comment]
Travis MacOS failure is unrelated, I think.

So I think this is done. does it [NeedsNEWS]?

**Comment by ararslan** [comment]
Using `WorkerProcessExitedException` in Distributed is potentially breaking, since code could be checking for `ProcessExitedException` in a `catch`.

If you consolidate `ProcessExitedException` and `WorkerProcessExitedException`, you could make the type's field a `Union{Vector{Process},Nothing}`, and define
```julia
ProcessExitedException() = ProcessExitedException(nothing)
```
for use in Distributed.

**Comment by oxinabox** [comment]
This is already breaking, since code could be checking for `ErrorException` in a `catch`.


The consolidation **could** be done. If you think that that kinda hackyness is preferable.
Though even then it would be breaking as code could be checking for `Distributions.ProcessExcitedException` in a `catch`.

(I would as an aside say there is an argument for `Union{Vector{Process}, Missing}`, the Process that exitted did exist, we just don't know what it was. But OTOH some people get ansty about `Missing` in a nonstatistical context.)

**Comment by ararslan** [comment]
> This is already breaking, since code could be checking for `ErrorException` in a `catch`.

Fair point, though I would guess (based on no actual data) that specifically looking for an `ErrorException` is pretty uncommon, except maybe in a `@test_throws`. We could check with a PkgEval run.

> If you think that that kinda hackyness is preferable.

I don't think it's hacky at all.

We could alternatively make Distributed fill in the field with the process information from the spawned Julia process, but that seems unnecessary since the information isn't used anywhere.

> Though even then it would be breaking as code could be checking for `Distributions.ProcessExcitedException` in a `catch`.

If `ProcessExitedException` is defined in Base, it is defined in all modules that load Base, including Distributed. For example, `using Distributed; Distributed.sin` works. So one can still use the Distributed qualification without issue.

File: base/process.jl [hunk 8]
```diff
@@ -701,9 +701,29 @@ with a code of 0). An exception is raised if the process cannot be started.
 """
 success(cmd::AbstractCmd) = success(_spawn(cmd))
 
+struct ProcessExitedException
+    procs::Union{Vector{Process},Nothing}
+    # ProcessExitedException(nothing) is allowed for Distributed stdlib compat
+end
+ProcessExitedException(proc::Process) = ProcessExitedException([proc])
```
Comment (ararslan) [inline base/process.jl]: You also need to define a constructor with no arguments in keeping with what was defined in Distributed.

Comment (ararslan) [reply base/process.jl]: > I guess in theory someone might have been constructing `ProcessExitedError()`
in their own code.

Yes, this

Comment (ararslan) [reply base/process.jl]: I'd prefer not to deprecate it anyway. `ProcessExitedException(nothing)` does not seem any more informative to me than `ProcessExitedException()`.

Comment (oxinabox) [reply base/process.jl]: Ok
Done

File: base/process.jl [hunk 9]
```diff
@@ -701,9 +701,30 @@ with a code of 0). An exception is raised if the process cannot be started.
...
 
+struct ProcessExitedException
+    procs::Union{Vector{Process},Nothing}
+    # ProcessExitedException(nothing) is allowed for Distributed stdlib compat
+end
+ProcessExitedException() = ProcessExitedException(nothing)
+ProcessExitedException(proc::Process) = ProcessExitedException([proc])
+
+function show(io::IO, err::ProcessExitedException)
+    if procs == nothing
```
Comment (ararslan) [inline base/process.jl]: `procs` isn't defined. Also it's better to use `===` when comparing against `nothing`.

File: test/spawn.jl [hunk 10]
```diff
@@ -423,13 +423,31 @@ end
...
+        show(iobuf, err)
+        errmsg = String(take!(iobuf))
+        @test occursin(string(failing_cmd), errmsg)
+    end
 end
 
+let
+    iobuf = IOBuffer()
+    show(iobuf, ProcessExitedException())
+    @test length(String(take!(iobuf))) > 0
```
Comment (ararslan) [inline test/spawn.jl]: This will pass regardless of whether a `show` method is even defined.

File: test/spawn.jl [hunk 11]
```diff
@@ -423,13 +423,31 @@ end
...
+failing_cmd = `$catcmd _doesnt_exist__111_`
+failing_pipeline = pipeline(failing_cmd, stderr=devnull) # make quiet for tests
+for testrun in (failing_pipeline, pipeline(failing_pipeline, failing_pipeline))
+    try
+        run(testrun)
+    catch err
+        @test err isa ProcessExitedException
+        iobuf = IOBuffer()
+        show(iobuf, err)
+        errmsg = String(take!(iobuf))
```
Comment (ararslan) [inline test/spawn.jl]: You can skip all of the buffer stuff and do `errmsg = sprint(show, err)`.

File: stdlib/Distributed/src/cluster.jl [hunk 12]
```diff
@@ -959,16 +959,6 @@ function _rmprocs(pids, waitfor)
...
 
 
-struct ProcessExitedException <: Exception end
-
-"""
-    ProcessExitedException()
-
-After a client Julia process has exited, further attempts to reference the dead child will
-throw this exception.
-"""
```
Comment (ararslan) [inline stdlib/Distributed/src/cluster.jl]: It would be good to add this docstring to the new definition in Base (updated accordingly) and make sure all of the docs reference the right thing.

**Comment by oxinabox** [comment]
Alright, I've mad my best shot a the docstring.

It is hard to document in an exceptioms docstring why it was thrown.
Espcially since the two reasons are fairly different.
I feel like that is more the job of the method that throws the exception.

File: base/process.jl [hunk 13]
```diff
@@ -661,6 +661,7 @@ Use [`pipeline`](@ref) to control I/O redirection.
 function run(cmds::AbstractCmd, args...; wait::Bool = true)
     if wait
         ps = _spawn(cmds, spawn_opts_inherit(args...)...)
+
```
Comment (ararslan) [inline base/process.jl]: Unrelated change

File: base/process.jl [hunk 14]
```diff
@@ -701,9 +702,43 @@ with a code of 0). An exception is raised if the process cannot be started.
...
+    ProcessExitedException
+
+Indicates problematic exit status of a process.
+
+Usages include:
+ 
+ - When running commands or pipelines, this is thrown to indicate
+ a nonzero exit code was returned (i.e. that the invoked process failed).
+ - In a [Distributed](@ref) multiprocessing workflow, this is thrown
+ when work is sent to a worker julia process that has exitted.
```
Comment (ararslan) [inline base/process.jl]: One t in exited

File: stdlib/Distributed/src/cluster.jl (hunk 12 above)
Comment (ararslan) [reply stdlib/Distributed/src/cluster.jl]: The language used here should just be copied verbatim into the new docstring.

**Review by ararslan — CHANGES_REQUESTED** [review]

**Comment by ararslan** [comment]
Still needs news, and it would be good to add `ProcessExitedException` to the documentation since it's exported.

File: base/process.jl [hunk 15]
```diff
@@ -780,9 +780,43 @@ with a code of 0). An exception is raised if the process cannot be started.
...
 success(cmd::AbstractCmd) = success(_spawn(cmd))
 
+
+"""
+    ProcessExitedException
+
+Indicates problematic exit status of a process.
+
+Usages include:
+
```
Comment (ararslan) [inline base/process.jl]: This line has the trailing whitespace that's making CI fail.

File: NEWS.md [hunk 16]
```diff
@@ -16,6 +16,11 @@ Multi-threading changes
 Language changes
 ----------------
 
+  * If a `pipeline` is specified with `append=true` set, but no redirection, an `ArgumentError`
```
Comment (ararslan) [inline NEWS.md]: The bullet points should be at column 1.

**Review by ararslan — DISMISSED** [review]

**Comment by oxinabox** [comment]
Done

**Comment by oxinabox** [comment]
Ok I finally worked out why I was breaking documentation. And fixed it.
How are things looking to merging this?

File: base/process.jl [hunk 17]
```diff
@@ -781,9 +781,43 @@ with a code of 0). An exception is raised if the process cannot be started.
...
+ when work is sent to a worker julia process that has exited.
+"""
+struct ProcessExitedException
+    procs::Union{Vector{Process},Nothing}
+    # ProcessExitedException(nothing) is allowed for Distributed stdlib compat
+end
+ProcessExitedException() = ProcessExitedException(nothing)
+ProcessExitedException(proc::Process) = ProcessExitedException([proc])
+
+function show(io::IO, err::ProcessExitedException)
```
[the remark this answers is not in the document]
Comment (oxinabox) [reply base/process.jl]: That sounds correct.

**Comment by GregPlowman** [comment]
Quite independently, I just put up a RFC #30733 involving `Distributed.ProcessExitedException`. In that PR I want to capture the worker pid in the exception.

Further to the discussion above, I'm wondering if two separate exceptions (one for command/pipelines and one for workers) would be better. Is combining them into one conflating two separate types? If it wasn't for backward compatibility, would that decision to combine them be made?

Using `Union{Vector{Process}, Nothing}` to accommodate different use cases does seem fishy, and somewhat un-julian.

Perhaps unnecessarily convoluted, but what about something like:
```
abstract type ProcessExitedException <: Exception end

struct CmdExitedException <: ProcessExitedException
    procs::Vector{Process}
end

struct WorkerExitedException <: ProcessExitedException 
    pid::Int
end
```
Or just using a different, non-conflicting name (say `CmdExitedException`) for the exception in this PR.
 
Having separate Exceptions would also be "safer" in the sense that `CmdExitedException` cannot be accidentally created without the `Process` argument. i.e. `CmdExitedException()` could be an error.

In any case, would it be possible to incorporate the intent of my PR into this one?

**Comment by oxinabox** [comment]
> Further to the discussion above, I'm wondering if two separate exceptions (one for command/pipelines and one for workers) would be better. Is combining them into one conflating two separate types?

> Using Union{Vector{Process}, Nothing} to accommodate different use cases does seem a bit fishy to me.


This is a fight I lost to @ararslan  3 months ago.

**Comment by oxinabox** [comment]
Can we review and merge this one,
and any further changes, re  #30733 can be discussed there,
after rebasing that ontop of this.

**Comment by oxinabox** [comment]
Bump?

**Comment by oxinabox** [comment]
bump?

**Comment by StefanKarpinski** [comment]
Since this passes tests and seems to be ready could someone with knowledge of the process stuff take a look and merge if it looks good? cc @Keno, @vtjnash

**Comment by oxinabox** [comment]
Bump? This has been waiting for review fore almost 4 weeks.

**Comment by StefanKarpinski** [comment]
I’ve marked it for triage but note that it appears to still need both docs and news.

**Comment by oxinabox** [comment]
It doesn't AFAICT And hasn't for a while.

Since https://github.com/JuliaLang/julia/pull/27900#issuecomment-450942486

**Comment by StefanKarpinski** [comment]
Ok, I've removed the outdated labels. Still marked for triage so it will get discussed.

**Comment by JeffBezanson** [comment]
:+1: to the overall concept here.

There is a distinction though: for a spawned process, we only throw an error if the process exits with a non-0 code. That sounds more like a `ProcessFailed` exception. But for `Distributed`, it's a problem if one of the worker processes exits at all, for any reason.

**Comment by vtjnash** [comment]
Seems like maybe some sort of general state / closed violation?

**Comment by oxinabox** [comment]
My plan was for them to be seperate, it was only made the same (IMO) to avoid a name clash.

I am sure i can revert some commits in this chain and seperate them out again.
Then rename the one from `spawn` to be `ProcessFailedException`

**Comment by oxinabox** [comment]
This is ready for rereview

File: base/process.jl [hunk 18]
```diff
@@ -781,9 +781,36 @@ with a code of 0). An exception is raised if the process cannot be started.
...
+"""
+    ProcessFailedException
+
+Indicates problematic exit status of a process.
+When running commands or pipelines, this is thrown to indicate
+a nonzero exit code was returned (i.e. that the invoked process failed).
+"""
+struct ProcessFailedException
+    procs::Vector{Process}
+
```
Comment (GregPlowman) [inline base/process.jl]: superfluous empty line?

File: base/process.jl [hunk 19]
```diff
@@ -781,9 +781,36 @@ with a code of 0). An exception is raised if the process cannot be started.
...
+end
+ProcessFailedException() = ProcessFailedException(nothing)
```
Comment (GregPlowman) [inline base/process.jl]: delete this constructor?

**Comment by oxinabox** [comment]
Bump?

**Comment by StefanKarpinski** [comment]
@vtjnash, please review and either provide requested changes or approve.

**Comment by oxinabox** [comment]
Another week passes, another bump made.
Thought is the arrow of time, memory never fades.

**Comment by StefanKarpinski** [comment]
If this isn't reviewed by next week, I'm going to merge it as is.

**Comment by GregPlowman** [comment]
Could I tempt any reviewers into a 2 for 1 offer?

Now that exceptions from failed spawned processes are distinct from worker exceptions, is it possible to also review #30733, which should now be independent from this PR

**Review by Keno — APPROVED** [review]
Triage approves, but @vtjnash has a small additional comment he'll make separately.

File: base/process.jl [hunk 20]
```diff
@@ -381,7 +381,7 @@ const SpawnIOs = Vector{Any} # convenience name for readability
             h === C_NULL     ? (0x00, UInt(0)) :
             h isa OS_HANDLE  ? (0x02, UInt(cconvert(@static(Sys.iswindows() ? Ptr{Cvoid} : Cint), h))) :
             h isa Ptr{Cvoid} ? (0x04, UInt(h)) :
-            error("invalid spawn handle $h from $io")
+            error("invalid spawn handle $h from $io") #TODO use a real error type here
```
Comment (vtjnash) [inline base/process.jl]: `error` is perfectly acceptable for errors, let's not add a TODO suggesting otherwise (this one is nearly an `@assert` on `rawhandle` giving the expected result)

Comment (oxinabox) [reply base/process.jl]: Earlier I had changed it to an `@assert`, somewhere in the history can find me changing it back.

Comment (oxinabox) [reply base/process.jl]: Done

**Review by vtjnash — APPROVED** [review]

**Comment by fredrikekre** [comment]
Looks like some `download` tests need to be fixed:
```
Worker 4 failed running test download:
Some tests did not pass: 8 passed, 2 failed, 0 errored, 0 broken.download: Test Failed at /tmp/julia/share/julia/test/download.jl:19
  Expression: download("http://httpbin.org/status/404", missing_file)
    Expected: ProcessFailedException
      Thrown: ErrorException
```

**Comment by oxinabox** [comment]
Bother, I fixed them before, but they must have broken during the rebase.

**Comment by oxinabox** [comment]
I can not reproduce the CI failures locally, download correctly gives `ProcessFailedErrors`.
not `ErrorExceptions`.

```
julia> invalid_host_file = joinpath(mktempdir(), "invalid_host")
"/var/folders/1r/4cjj1xk90qbb6njj7dn_cmt40000gn/T/tmpaODPDn/invalid_host"

julia> ProcessFailedException download("http://192.0.2.1", invalid_host_file)
ERROR: syntax: extra token "download" after end of expression

julia> download("http://192.0.2.1", invalid_host_file)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:11 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:12 --:--:--     0

  0     0    0     0    0     0      0      0 --:--:--  0:01:15 --:--:--     0curl: (7) Failed to connect to 192.0.2.1 port 80: Operation timed out
ERROR: failed process: Process(`/usr/bin/curl -g -L -f -o /var/folders/1r/4cjj1xk90qbb6njj7dn_cmt40000gn/T/tmpaODPDn/invalid_host http://192.0.2.1`, ProcessExited(7)) [7]

Stacktrace:
 [1] pipeline_error at ./process.jl:811 [inlined]
 [2] #run#519(::Bool, ::typeof(run), ::Cmd) at ./process.jl:727
 [3] run at ./process.jl:725 [inlined]
 [4] download(::String, ::String) at ./download.jl:27
 [5] top-level scope at REPL[8]:1

julia>

julia> download("a")
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: a
ERROR: failed process: Process(`/usr/bin/curl -g -L -f -o /var/folders/1r/4cjj1xk90qbb6njj7dn_cmt40000gn/T/juliacOKV9q a`, ProcessExited(6)) [6]

Stacktrace:
 [1] pipeline_error at ./process.jl:811 [inlined]
 [2] #run#519(::Bool, ::typeof(run), ::Cmd) at ./process.jl:727
 [3] run at ./process.jl:725 [inlined]
 [4] download(::String, ::String) at ./download.jl:27
 [5] download(::String) at ./download.jl:47
 [6] top-level scope at REPL[9]:1
```

So I just did a rebase and force push to rekickoff the CI.

**Comment by oxinabox** [comment]
Ok, I think I found it.
Extra error handling had been added to the `download` to give some extra information.
Which was set to throw with `error`, and thus changed the type of the exception from what was prevously thrown.
To keep it so that wether  or not we have extra information to add, the same Exception type is thrown,
I changed it to instead log that extra info to `@error`.

**Comment by oxinabox** [comment]
The Travis and AppVeyor failures are unrelated.
I can kick it a few more times, or if we are satisfied we can merge?

**Comment by oxinabox** [comment]
Can this go in now? Before the feature freeze.
I've rebased through multiple releases now and it is annoying to do, and the code changes under me.

Assuming noone has a dread hatred of using the `@error` logger for reporting supplementary error information.
Then this is done and approved.

**Comment by StefanKarpinski** [comment]
I'm going to merge as-is tomorrow if this doesn't get a review...

**Comment by JeffBezanson** [comment]
This has had multiple approved reviews for a long time.

File: base/download.jl [hunk 21]
```diff
@@ -39,8 +39,9 @@ function download_curl(curl_exe::AbstractString, url::AbstractString, filename::
     err = PipeBuffer()
     process = run(pipeline(`$curl_exe -s -S -g -L -f -o $filename $url`, stderr=err), wait=false)
     if !success(process)
-        stderr = readline(err)
-        error(stderr)
+        error_msg = readline(err)
+        @error "Download Failed" error_msg
```
Comment (musm) [inline base/download.jl]: this should've been a string interpolation, fixed in https://github.com/JuliaLang/julia/pull/31620
